blog/2023-04-18-magazine-003.md

```diff
@@ -0,0 +1,325 @@
+
+```js
+bring cloud;
+bring "aws-cdk-lib" as aws;
+
+let queue = new cloud.Queue();
+let bucket = new cloud.Bucket();
+
+queue.add_consumer(inflight (message: str) => {
+  bucket.put("latest_message.txt", message);
+});
+
+let func = new cloud.Function(inflight () => {
+  queue.push("hello world!");
+});
+
+let table = new aws.aws_dynamodb.Table();
+```
+
+As you can see, several resources from our batteries-included SDK are available in the AWS CDK target. Whenever you use these resources, your app can be deployed **both** to AWS using CloudFormation or to the Wing simulator.
+
+Since we are compiling to AWS CDK, we are also able to use any constructs from `aws-cdk-lib` directly in Wing.
+Since these imported resources don't know anything about Wing's "inflights", IAM permissions for these resources can't be inferred automatically by the compiler, and they do anything in the Wing simulator.
+But it's still possible to define permissions explicitly using AWS CDK APIs, and in the future we're planning to add mechanisms to specify how foreign resources behaves in the simulator.
+
+Try it out and let us know what you think! Huge thanks again to [Marcio](https://www.linkedin.com/in/marcio-almeida-25891669/) for adding support for this feature!
+
+### Extern functions - call out to JavaScript straight from Wing
+
+A common ask from developers we talk to (and our own team) has been -- if Wing is able to compile applications down to JavaScript (and it is!), how can we reuse existing JavaScript code and libraries in Wing apps?
+
+The answer is **extern functions**. Extern functions let you specify a function is not implemented in Wing, but by a piece of external JavaScript. Here's an example - let's say we have a file called `helpers.js` that contains the following code that generates unique IDs for us:
+
+```js
+const uuid = require("uuid");
+
+exports.make_id = async function() {
+    return uuid.v4();
+}
+```
+
+(In order for this to work, you'll also need to run `npm install uuid` in the same directory as `helpers.js`).
+
+We can then use this function in our Wing app as if it were a regular Wing function:
+
+```js
+bring cloud;
+
+resource Utils {
+  init() {}
+  // define the extern function here...
+  extern "./helpers.js" inflight make_id(): str;
+}
+
+let bucket = new cloud.Bucket();
+let utils = new Utils();
+
+new cloud.Function(inflight () => {
+  // ...and use it inflight here
+  let id = utils.make_id();
+  bucket.put(id, "hello world!");
+}) as "test";
+```
+
+If you simulate the app locally using the Wing Console, your development machine will call the `make_id` function in `helpers.js` and use the copy of `uuid` you already have installed when simulating `cloud.Function`.
+
+But if you compile the app to the `tf-aws` target, the Wing will bundle all of your code -- including the `uuid` dependency -- into a single AWS Lambda function, so that the code will work the same way in the cloud as it does in the simulator.
```

Comment:
```suggestion
But if you compile the app to the `tf-aws` target, Wing will bundle all of your code -- including the `uuid` dependency -- into a single AWS Lambda function, so that the code will work the same way in the cloud as it does in the simulator.
```